src/persist-txn/src/metrics.rs

```diff
@@ -0,0 +1,191 @@
+                var_labels: ["op"],
+            )),
+            retry_count: registry.register(metric!(
+                name: "mz_txn_op_retry_count",
+                help: "count of times a txn operation retried",
+                var_labels: ["op"],
+            )),
+            succeeded_count: registry.register(metric!(
+                name: "mz_txn_op_succeeded_count",
+                help: "ount of times a txn operation succeeded",
+                var_labels: ["op"],
+            )),
+            errored_count: registry.register(metric!(
+                name: "mz_txn_op_errored_count",
+                help: "ount of times a txn operation errored",
+                var_labels: ["op"],
+            )),
+            duration_seconds: registry.register(metric!(
+                name: "mz_txn_op_duration_seconds",
+                help: "time spent running a txn operation",
+                var_labels: ["op"],
+            )),
+        }
+    }
+
+    fn fallible(&self, name: &str) -> FallibleOpMetrics {
+        FallibleOpMetrics {
+            started_count: self.started_count.with_label_values(&[name]),
+            retry_count: self.retry_count.with_label_values(&[name]),
+            succeeded_count: self.succeeded_count.with_label_values(&[name]),
+            errored_count: self.errored_count.with_label_values(&[name]),
+            duration_seconds: self.duration_seconds.with_label_values(&[name]),
+        }
+    }
+
+    fn infallible(&self, name: &str) -> InfallibleOpMetrics {
+        InfallibleOpMetrics {
+            started_count: self.started_count.with_label_values(&[name]),
+            succeeded_count: self.succeeded_count.with_label_values(&[name]),
+            duration_seconds: self.duration_seconds.with_label_values(&[name]),
+        }
+    }
+}
+
+pub(crate) struct FallibleOpMetrics {
+    started_count: IntCounter,
+    pub(crate) retry_count: IntCounter,
+    succeeded_count: IntCounter,
+    errored_count: IntCounter,
+    duration_seconds: Counter,
+}
+
+impl FallibleOpMetrics {
+    pub(crate) async fn run<R, E>(&self, op: impl Future<Output = Result<R, E>>) -> Result<R, E> {
+        let start = Instant::now();
+        self.started_count.inc();
+        let res = op.await;
+        match res {
+            Ok(_) => self.succeeded_count.inc(),
+            Err(_) => self.errored_count.inc(),
+        }
+        self.duration_seconds.inc_by(start.elapsed().as_secs_f64());
+        res
+    }
+}
+
+pub(crate) struct InfallibleOpMetrics {
+    started_count: IntCounter,
+    succeeded_count: IntCounter,
+    duration_seconds: Counter,
+}
+
+impl InfallibleOpMetrics {
+    pub(crate) async fn run<R>(&self, op: impl Future<Output = R>) -> R {
+        let start = Instant::now();
+        self.started_count.inc();
+        let res = op.await;
+        self.succeeded_count.inc();
+        self.duration_seconds.inc_by(start.elapsed().as_secs_f64());
+        res
+    }
+}
```

Comment:
I _think_ `mz_ore` already has some of this functionality. Not sure if it's exactly what you want though: https://github.com/MaterializeInc/materialize/blob/068feb635e8ef16669ec000accf88b5099556a2b/src/ore/src/metrics.rs#L411-L745

Reply:
Yeah,`.wall_time` gets us the execution time, but this helper also covers counting how many times it's called, etc. Not sure reusing the ore stuff gets us a ton